packages/gatsby-source-contentful/src/backreferences.js

```diff
@@ -0,0 +1,143 @@
+// @ts-check
+import { hasFeature } from "gatsby-plugin-utils/index"
+import { getDataStore } from "gatsby/dist/datastore"
+import { untilNextEventLoopTick } from "./utils"
+
+// Array of all existing Contentful nodes. Make it global and incrementally update it because it's hella slow to recreate this on every data update for large sites.
+export const existingNodes = new Map()
+
+let allNodesLoopCount = 0
+export const is = {
+  firstSourceNodesCallOfCurrentNodeProcess: true,
+}
+
+export async function getExistingCachedNodes({
+  actions,
+  getNode,
+  pluginConfig,
+}) {
+  const { enableStatefulSourceNodes, touchNode } = actions
+
+  const hasStatefulSourceNodes = hasFeature(`stateful-source-nodes`)
+
+  if (hasStatefulSourceNodes) {
+    enableStatefulSourceNodes()
+  }
```

Comment:
That's fair, just moved it! I put it in the other place so it was near the code that calls touchNode, but I'm good with either :)

Reply:
Also, in a future breaking release I think we should make `enableStatefulSourceNodes` the default and remove `touchNode` entirely. We could move that code out into `plugin-utils` so the plugins that need it can check themselves for stale nodes. 
I think stale checking makes sense for fs sources but otherwise encourages source plugins to be built in a non performant way.